src/rgw/rgw_tracer.h

```diff
@@ -0,0 +1,77 @@
+#ifndef TRACER_WRAPPER_H
+#define TRACER_WRAPPER_H
+
+#define SIGNED_RIGHT_SHIFT_IS 1
+#define ARITHMETIC_RIGHT_SHIFT 1
+#include <jaegertracing/Tracer.h>
+#include <yaml-cpp/yaml.h>
+#include "common/config.h"
+#include "common/ceph_context.h"
+
+
+namespace RGW_Tracing
+{
+    class TracerConfiguration : public md_config_obs_t {
+        private:
+            bool enabled = true;
+
+        public:
+            TracerConfiguration();
+            bool is_enabled() const;
+            const char** get_tracked_conf_keys() const override;
+            void handle_conf_change(const ConfigProxy& conf,
+                                    const std::set<std::string>& changed) override;
+    };
+
+    
+
+    class Span {
+        private:
+            std::unique_ptr<opentracing::Span> span;
+        public:
+            Span(std::unique_ptr<opentracing::Span> _span);
+            ~Span();
+
+            std::string get_baggage(std::string key) const;
+            opentracing::SpanReference get_reference() const;
+        
+            void set_tag(std::string key,std::string value);
+            void set_log(std::string key,std::string value);
+            /* set <key,value> pair which will be propagated to child spans */
+            void set_baggage(std::string key,std::string value);
+            
+    };
+
+
+    class Tracer {
+    private:
+        std::shared_ptr<opentracing::Tracer> open_tracer;
+    
+    public:
+        const static std::shared_ptr<opentracing::Tracer>   noop_tracer;
```

Comment:
hmm.  when is the shared_ptr accessed?

Reply:
why is this public?

Reply:
yuval and I discussed the noop_tracer in a 1x1 meeting;  as Yuval describes it, noop_tracer traces nothing, so why is it a shared_ptr (which is similar to a lock, so costs something to access)?  could we rework this slightly?

Reply:
it should be possible to wrap around the libraries "make()" functions and convert the `shared_ptr` to a `unique_ptr`.
e.g.
```cpp
std::unique_ptr<opentracing::Tracer> MakeNoopTracer() {
  return std::unique_ptr<opentracing::Tracer>(new opentracing::Tracer(*opentracing::MakeNoopTracer()));
}

std::unique_ptr<opentracing::Tracer> make(const jConfig& conf) {
  return std::unique_ptr<opentracing::Tracer>(new opentracing::Tracer(*opentracing::make(conf)));
}
```


however, this is mainly for the cleanliness of the code (and won't have perf implications), since we don't copy the `shared_ptr` around, we only pass it by reference. and dereferencing the `shared_ptr` should be a lockfree action

Reply:
you are right. 
@yuvalif  would it be better to make those convert functions static?

Reply:
I've looked in the implementation of the jaeger, and the tracer ptr is also a member of each span object. this is why jaeger/open-tracing uses a shared ptr. 
so it can't be done using unique ptr.

Reply:
ok. so lets keep it with `shared_ptr`